**Goal of the project**


The ability to predict that a particular customer is at a high risk of churning, while there is still time to do something about it, represents a huge additional potential revenue source for every online business. Besides the direct loss of revenue that results from a customer abandoning the business, the costs of initially acquiring that customer may not have already been covered by the customer’s spending to date. (In other words, acquiring that customer may have actually been a losing investment.) Furthermore, it is always more difficult and expensive to acquire a new customer than it is to retain a current paying customer.

In this project, we’ll build a contractual churn model for contractual settings.

**Load the package**

In [207]:
# Importing library
import pandas as pd

**Load the data**

For this project I’ve used the [Iranian Churn](https://www.kaggle.com/datasets/royjafari/customer-churn) dataset from the UCI Machine Learning Repository. This is data from a telecoms provider, so we’ll be using it here to create a contractual churn model.

In [208]:
# Load dataset
df = pd.read_csv('../input/customer-churn/Customer Churn.csv')

In [209]:
# Rename Pandas columns to lower case
df.columns = df.columns.str.lower()

In [210]:
# Examine the data
df.head()

call  failure  complains  subscription  length  charge  amount  \
0              8          0                    38               0   
1              0          0                    39               0   
2             10          0                    37               0   
3             10          0                    38               0   
4              3          0                    38               0   

   seconds of use  frequency of use  frequency of sms  \
0            4370                71                 5   
1             318                 5                 7   
2            2453                60               359   
3            4198                66                 1   
4            2393                58                 2   

   distinct called numbers  age group  tariff plan  status  age  \
0                       17          3            1       1   30   
1                        4          2            1       2   25   
2                       24          3            1       1   30   
3                       35          1            1       1   15   
4                       33          1            1       1   15   

   customer value         fn        fp  churn  
0         197.640   177.8760   69.7640      0  
1          46.035    41.4315   60.0000      0  
2        1536.520  1382.8680  203.6520      0  
3         240.020   216.0180   74.0020      0  
4         145.805   131.2245   64.5805      0

In [211]:
# Overview of all variables, their datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   call  failure            3150 non-null   int64  
 1   complains                3150 non-null   int64  
 2   subscription  length     3150 non-null   int64  
 3   charge  amount           3150 non-null   int64  
 4   seconds of use           3150 non-null   int64  
 5   frequency of use         3150 non-null   int64  
 6   frequency of sms         3150 non-null   int64  
 7   distinct called numbers  3150 non-null   int64  
 8   age group                3150 non-null   int64  
 9   tariff plan              3150 non-null   int64  
 10  status                   3150 non-null   int64  
 11  age                      3150 non-null   int64  
 12  customer value           3150 non-null   float64
 13  fn                       3150 non-null   float64
 14  fp                      

**Examine the data**

As we’ll see from examining the value_counts( ) of the target variable column, this dataset is imbalanced. The positive class (customers who churned) comprise about 16% of the dataset. This is the norm for customer churn datasets in non contractual and contractual settings, but does introduce some challenges we need to handle, otherwise the model may fail to make accurate predictions.

In [212]:
df['churn'].value_counts()

0    2655
1     495
Name: churn, dtype: int64

**Check for missing values**

This dataset has probably been pre-cleansed, as df.isnull( ).sum( ) shows us that we don’t have any null values to deal with at all, which will save us a step later.

In [213]:
# Check for missing values
df.isnull().sum()

call  failure              0
complains                  0
subscription  length       0
charge  amount             0
seconds of use             0
frequency of use           0
frequency of sms           0
distinct called numbers    0
age group                  0
tariff plan                0
status                     0
age                        0
customer value             0
fn                         0
fp                         0
churn                      0
dtype: int64

**Create the training and test data**

Now we have prepared our feature set, we need to define X and y. The X feature set will include all the features we created above, minus the target variable, so we don’t give the model the answer. The y data will comprise the target variable alone.

In [214]:
X = df.drop('churn', axis = 1) 

In [215]:
y = df['churn']

We’ll now randomly split the data into a training and test dataset using the train_test_split( ) function. We’ll assign 30% of the data to the test group, which will be held out of training, and we’ll use the stratify option to ensure the target variable is present in equal proportions in each group. The random_state value ensures we get reproducible results each time we run the code.

In [216]:
from sklearn.model_selection import train_test_split

In [217]:
# Isolate X and y variables, and perform train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3, random_state = 42, stratify = y)

**Create a model pipeline**

Next, we’ll create a model pipeline. This will handle the encoding of our data using the ColumnTransformer( ) feature.

I’ve also added some basic feature selection via SelectKBest( ), and have used the Synthetic Minority Oversampling Technique and Edited Nearest Neighbor (SMOTE-ENN) method to better handle class imbalance. 

In [218]:
import time
from sklearn.compose import ColumnTransformer

from imblearn.pipeline import Pipeline as imbpipeline
from imblearn.combine import SMOTEENN
from sklearn.preprocessing import RobustScaler
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import mutual_info_classif
from sklearn.model_selection import cross_val_score
from sklearn.metrics import roc_auc_score

In [219]:
def get_pipeline(X, model):

    numeric_columns = list(X.select_dtypes(exclude = ['object']).columns.values.tolist())    
    
    preprocessor = ColumnTransformer(transformers = [('numeric', 'passthrough', numeric_columns)], remainder = 'passthrough')

    bundled_pipeline = imbpipeline(steps = [('preprocessor', preprocessor),
                                            ('smote', SMOTEENN(random_state = 42)),
                                            ('scaler', RobustScaler()),
                                            ('feature_selection', SelectKBest(score_func = mutual_info_classif, k = 13)),
                                            ('model', model)])
    
    return bundled_pipeline

**Select the best model**

Rather than simply selecting a single model, or repeating our code manually on a range of models, we can create another function to automatically test a wide range of possible models to determine the best one for our needs. To do this we first create a dictionary containing some a selection of base classifiers.

We will create a Pandas dataframe into which we will store the data. Then we will loop over each of the models, fit it using the X_train and y_train data, then generate predictions from X_test and calculate the mean ROC/AUC score from 5 rounds of cross-validation. That will give us the ROC/AUC score for the X_test data, plus the average ROC/AUC score for the training dataset.

In [220]:
from xgboost import XGBClassifier, XGBRFClassifier
from lightgbm import LGBMClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, ExtraTreesClassifier, GradientBoostingClassifier
from sklearn.ensemble import BaggingClassifier, AdaBoostClassifier, HistGradientBoostingClassifier
from catboost import CatBoostClassifier

In [221]:
def select_model(X, y, pipeline = None):

  classifiers = {}
  classifiers.update({'XGBClassifier': XGBClassifier(random_state = 42)})
  classifiers.update({'XGBRFClassifier': XGBRFClassifier(random_state = 42)})
  classifiers.update({'LGBMClassifier': LGBMClassifier(random_state = 42)})
  classifiers.update({'DecisionTreeClassifier': DecisionTreeClassifier(random_state = 42)})
  classifiers.update({'RandomForestClassifier': RandomForestClassifier(random_state = 42)})
  classifiers.update({'ExtraTreesClassifier': ExtraTreesClassifier(random_state = 42)})
  classifiers.update({'GradientBoostingClassifier': GradientBoostingClassifier(random_state = 42)})    
  classifiers.update({'BaggingClassifier': BaggingClassifier(random_state = 42)})
  classifiers.update({'AdaBoostClassifier': AdaBoostClassifier(random_state = 42)})
  classifiers.update({'HistGradientBoostingClassifier': HistGradientBoostingClassifier(random_state = 42)})
  classifiers.update({'CatBoostClassifier': CatBoostClassifier(silent = True, random_state = 42)})

  df_models = pd.DataFrame(columns = ['model', 'run_time', 'roc_auc_cv', 'roc_auc'])

  for key in classifiers:

      print('*', key)

      start_time = time.time()
      
      pipeline = get_pipeline(X_train, classifiers[key])

      cv = cross_val_score(pipeline, X, y, cv = 5, scoring = 'roc_auc', n_jobs = -1)

      pipeline.fit(X_train, y_train)
      y_pred = pipeline.predict(X_test)

      row = {'model': key,
             'run_time': format(round((time.time() - start_time) / 60, 2)),
             'roc_auc_cv': cv.mean(),
             'roc_auc': roc_auc_score(y_test, y_pred)}

      df_models = df_models.append(row, ignore_index = True)

  df_models = df_models.sort_values(by = 'roc_auc', ascending = False)
      
  return df_models

The top performer was HistGradientBoostingClassifier( ), which generated a ROC/AUC score of 0.942. We’ll select this model as our best one and examine the results in a bit more detail to see how well it works.

In [222]:
models = select_model(X_train, y_train)

* XGBClassifier
* XGBRFClassifier
* LGBMClassifier
* DecisionTreeClassifier
* RandomForestClassifier
* ExtraTreesClassifier
* GradientBoostingClassifier
* BaggingClassifier
* AdaBoostClassifier
* HistGradientBoostingClassifier
* CatBoostClassifier


In [223]:
models.head(10)

model run_time  roc_auc_cv   roc_auc
9   HistGradientBoostingClassifier     0.05    0.966518  0.942233
2                   LGBMClassifier     0.03    0.966288  0.936972
10              CatBoostClassifier     0.53    0.966784  0.935090
0                    XGBClassifier     0.05    0.964216  0.932725
7                BaggingClassifier     0.02    0.949656  0.931326
4           RandomForestClassifier     0.05    0.959437  0.930940
5             ExtraTreesClassifier     0.03    0.966763  0.925824
3           DecisionTreeClassifier     0.01    0.880014  0.914917
6       GradientBoostingClassifier     0.07    0.955135  0.913133
8               AdaBoostClassifier     0.03    0.947986  0.903867

**Examine the performance of the best model**

By re-running the get_pipeline( ) function on our selected model, the HistGradientBoostingClassifier( ), we can generate predictions and assess their accuracy on the test data.

In [224]:
selected_model = HistGradientBoostingClassifier(random_state = 42)
bundled_pipeline = get_pipeline(X_train, selected_model)
bundled_pipeline.fit(X_train, y_train)
y_pred = bundled_pipeline.predict(X_test)

**Examine the predictions**

The classification_report( ) function is also worth running on the data. For each class, this effectively shows how good the model was at predicting each class. The closer to 1 the better the performance.

In [225]:
from sklearn.metrics import classification_report

In [226]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.94      0.96       797
           1       0.74      0.95      0.83       148

    accuracy                           0.94       945
   macro avg       0.87      0.94      0.90       945
weighted avg       0.95      0.94      0.94       945



Let’s unpack those results a little bit…

**Recall**


A churn class recall of 0.95 means that the model was able to catch 95% of the actual churn cases. This is the measure we really care about, because we want to miss as few of the true churn cases as possible.

**Precision**

Precision of the churn class measures how often the model catches an actual churn case, while also factoring in how often it misclassifies a non-churn case as a churn case. In this case, a churn precision of 0.74 is not a problem because there are no significant consequences of identifying a customer as a churn risk when she isn’t.

**F1 score**

The F1 score is the harmonic mean of precision and recall. It helps give us a balanced idea of how the model is performing on the churn class. In this case a churn class F1 score of 0.83 is pretty good. 